src/license_manager_cli_ops.py

```diff
@@ -0,0 +1,202 @@
+        ]
+        subprocess.call(upgrade_pip_cmd)
+
+        # Install license-manager-cli package
+        pip_install_cmd = [
+            self._VENV_PYTHON,
+            "-m",
+            "pip",
+            "install",
+            self._PACKAGE_NAME,
+        ]
+        logger.debug(f"## Running: {pip_install_cmd}")
+        try:
+            out = subprocess.check_output(pip_install_cmd, env={}).decode().strip()
+            logger.debug("license-manager-cli installed")
+            logger.debug(f"## pip install output: {out}")
+        except Exception as e:
+            logger.error(f"Error installing license-manager-cli: {e}")
+            raise Exception("license-manager-cli not installed.")
+
+        # Create the lm-cli script file
+        self.configure_bin_script()
+
+        # Create the etc/default/lm-cli file
+        self.configure_etc_default()
+
+
+    def setup_cache_dir(self):
+        """Set up cache dir."""
+
+        # Delete cache dir if it already exists
+        if self._CACHE_DIR.exists():
+            rmtree(self._CACHE_DIR, ignore_errors=True)
+        # Create a clean cache dir
+        self._CACHE_DIR.mkdir(parents=True)
+        chown(self._CACHE_DIR.as_posix(), self._SLURM_USER, self._SLURM_GROUP)
+        self._CACHE_DIR.chmod(0o700)
+
+
+    def setup_log_dir(self):
+        """Set up log dir."""
+
+        # Create log dir
+        if not self._LOG_DIR.exists():
+            self._LOG_DIR.mkdir(parents=True)
+        chown(self._LOG_DIR.as_posix(), self._SLURM_USER, self._SLURM_GROUP)
+        self._LOG_DIR.chmod(0o700)
+
+
+    def upgrade(self, version: str):
+        """Upgrade license-manager-cli package to specified version."""
+
+        pip_install_cmd = [
+            self._VENV_PYTHON,
+            "-m",
+            "pip",
+            "--upgrade",
+            f"{self._PACKAGE_NAME}=={version}",
+        ]
+
+        out = subprocess.check_output(pip_install_cmd).decode().strip()
+        if "Successfully installed" not in out:
+            logger.error("Trouble upgrading license-manager-cli, please debug")
+            raise Exception("license-manager-cli not upgraded.")
+        else:
+            logger.debug("license-manager-cli upgraded")
+
+        # Clear cache dir after upgrade to avoid stale data
+        self.setup_cache_dir()
+
+
+    def configure_etc_default(self):
+        """Create the default env file with the charm's configurations."""
+        charm_config = self._charm.model.config
+        license_manager_backend_url = charm_config.get("license-manager-backend-url")
+        oidc_domain = charm_config.get("oidc-domain")
+        oidc_audience = charm_config.get("oidc-audience")
+        oidc_client_id = charm_config.get("oidc-client-id")
+
+        ctxt = {
+            "license_manager_backend_url": license_manager_backend_url,
+            "oidc_domain": oidc_domain,
+            "oidc_audience": oidc_audience,
+            "oidc_client_id": oidc_client_id,
+        }
```

Comment:
a smarter alternative with no need to define manually the items:

```python
ctxt = dict(map(lambda item: {item[0].replace("-", "_"), item[1]}, charm_config.keys()))
```

Reply:
I prefer map, but anyway, both are better indeed

Reply:
I'll change it to use dict comprehension.